src/app/pages/TopicPage/Pagination/buildBlocks.js

```diff
@@ -0,0 +1,185 @@
+};
+
+// After setting the visibility of all the pages we want to show, we can remove the others
+const pruneInvisible = () => {
+  state.result = state.result.filter((page, index) => {
+    if (!page.visibility) {
+      // if an element is being filtered out, we need to remember we did this
+      // this is so we can display an ellipsis in this position
+      if (index < state.activePageIndex) {
+        state.pagesTruncatedOnLeft = true;
+      } else {
+        state.pagesTruncatedOnRight = true;
+      }
+      return false;
+    }
+    return true;
+  });
+};
+
+// Determine the devices that an ellipsis is displayed on
+const getEllipsisVisibility = side => {
+  // If we pruned some pages on this side, we display an ellipsis on all devices
+  const wasTruncated =
+    side === 'left' ? state.pagesTruncatedOnLeft : state.pagesTruncatedOnRight;
+  if (wasTruncated) return VISIBILITY.ALL;
+
+  // Otherwise, the ellipsis visibility is based on the visibility of the page on that edge
+  // eg, if page 2 is visible on all devices, there will never be an ellipsis on the left
+  // if it is only visible on tablets and up, there will be an ellipsis on mobile
+  const highestVisibility =
+    side === 'left'
+      ? state.highestVisibilityOnLeft
+      : state.highestVisibilityOnRight;
+
+  if (!highestVisibility || highestVisibility === VISIBILITY.ALL) return null;
+  if (highestVisibility === VISIBILITY.TABLET_UP) return VISIBILITY.MOBILE_ONLY;
+  if (highestVisibility === VISIBILITY.DESKTOP_ONLY)
+    return VISIBILITY.TABLET_DOWN;
+  return null;
+};
+
+// Conditionally adding the ellipsis blocks to our return value
+const insertEllipsis = () => {
+  const leftEllipsisVisibility = getEllipsisVisibility('left');
+  const rightEllipsisVisibility = getEllipsisVisibility('right');
+  if (leftEllipsisVisibility) {
+    state.result.splice(1, 0, {
+      type: TYPE.ELLIPSIS,
+      visibility: leftEllipsisVisibility,
+    });
+  }
+
+  if (rightEllipsisVisibility) {
+    state.result.splice(state.result.length - 1, 0, {
+      type: TYPE.ELLIPSIS,
+      visibility: rightEllipsisVisibility,
+    });
+  }
+};
+
+// We display left and right arrows on all devices
+const insertArrows = () => {
+  state.result.unshift({
+    type: TYPE.LEFT_ARROW,
+    visibility: VISIBILITY.ALL,
+    // The left arrow is disabled if the user is already on the first page
+    availability:
+      state.activePage === 1
+        ? AVAILABILITY.UNAVAILABLE
+        : AVAILABILITY.AVAILABLE,
+  });
+  state.result.push({
+    type: TYPE.RIGHT_ARROW,
+    visibility: VISIBILITY.ALL,
+    // The right arrow is disabled if the user is already on the last page
+    availability:
+      state.activePage === state.pageCount
+        ? AVAILABILITY.UNAVAILABLE
+        : AVAILABILITY.AVAILABLE,
+  });
+};
+
+const addKeys = () => {
+  state.result.forEach((result, i) => {
+    // eslint-disable-next-line no-param-reassign
+    result.key = i;
+  });
```

Comment:
```suggestion
  state.result = state.result.map((result, i) => {
    return { ...result, key: i };
  });
```
Think this will remove the need for the linting comment and make this more functional

Reply:
👍  this is a bigger problem since moving `state` to being passed into each function

really don't think it's worth the added code complexity to try to make all of these helper functions pure

they're all private - the default export is pure